drivers/shared/executor/exec_utils.go

```diff
@@ -0,0 +1,295 @@
+	handleStdout(e.logger, stdoutPr, &wg, send, errCh)
+	handleStderr(e.logger, stderrPr, &wg, send, errCh)
+
+	ps, err := e.processWait()
+
+	// force close streams to close out the stream copying goroutines
+	stdinPr.Close()
+	stdoutPw.Close()
+	stderrPw.Close()
+
+	// wait until we get all process output
+	wg.Wait()
+
+	// wait to flush out output
+	stream.Send(cmdExitResult(ps, err))
+
+	select {
+	case cerr := <-errCh:
+		return cerr
+	default:
+		return nil
+	}
+}
+func cmdExitResult(ps *os.ProcessState, err error) *drivers.ExecTaskStreamingResponseMsg {
+	exitCode := -1
+
+	if ps == nil {
+		if ee, ok := err.(*exec.ExitError); ok {
+			ps = ee.ProcessState
+		}
+	}
+
+	if ps == nil {
+		exitCode = -2
+	} else if status, ok := ps.Sys().(syscall.WaitStatus); ok {
+		exitCode = status.ExitStatus()
+		if status.Signaled() {
+			const exitSignalBase = 128
+			signal := int(status.Signal())
+			exitCode = exitSignalBase + signal
+		}
+	}
+
+	return &drivers.ExecTaskStreamingResponseMsg{
+		Exited: true,
+		Result: &dproto.ExitResult{
+			ExitCode: int32(exitCode),
+		},
+	}
+}
+
+func handleStdin(logger hclog.Logger, stdin io.WriteCloser, stream drivers.ExecTaskStream, errCh chan<- error) {
+	go func() {
+		for {
+			m, err := stream.Recv()
+			if isClosedError(err) {
+				return
+			} else if err != nil {
+				errCh <- err
+				return
+			}
+
+			if m.Stdin != nil {
+				if len(m.Stdin.Data) != 0 {
+					_, err := stdin.Write(m.Stdin.Data)
+					if err != nil {
+						errCh <- err
+						return
+					}
+				}
+				if m.Stdin.Close {
+					stdin.Close()
+				}
+			} else if m.TtySize != nil {
+				err := setTTYSize(stdin, m.TtySize.Height, m.TtySize.Width)
+				if err != nil {
+					errCh <- fmt.Errorf("failed to resize tty: %v", err)
+					return
+				}
+			} else {
+				// ignore heartbeats
+			}
+		}
+	}()
+}
+
+func handleStdout(logger hclog.Logger, reader io.Reader, wg *sync.WaitGroup, send func(*drivers.ExecTaskStreamingResponseMsg) error, errCh chan<- error) {
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		for {
+			buf := make([]byte, 4096)
+			n, err := reader.Read(buf)
+			// always send output first if we read something
+			if n > 0 {
+				if err := send(&drivers.ExecTaskStreamingResponseMsg{
+					Stdout: &dproto.ExecTaskStreamingIOOperation{
+						Data: buf[:n],
+					},
+				}); err != nil {
+					errCh <- err
+					return
+				}
+			}
+
+			// then process error
+			if isClosedError(err) {
+				if err := send(&drivers.ExecTaskStreamingResponseMsg{
+					Stdout: &dproto.ExecTaskStreamingIOOperation{
+						Close: true,
+					},
+				}); err != nil {
+					errCh <- err
+					return
+				}
+				return
+			} else if err != nil {
+				errCh <- err
+				return
+			}
+
+		}
+	}()
+}
+
+func handleStderr(logger hclog.Logger, reader io.Reader, wg *sync.WaitGroup, send func(*drivers.ExecTaskStreamingResponseMsg) error, errCh chan<- error) {
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		for {
+			buf := make([]byte, 4096)
```

Comment:
I thought in #5632 you often reused buffers because you were careful to make `send()` not hold onto the buffer after returning? Is that not the case in this code? It would be nice to try to keep the approaches consistent or at least comment carefully when they diverge.

Reply:
Good catch - I missed these during the refactor - will make buf outside the `for` loop